x-pack/plugins/security_solution/server/lib/detection_engine/signals/utils.ts

```diff
@@ -439,3 +446,125 @@ export const getSignalTimeTuples = ({
   );
   return totalToFromTuples;
 };
+
+/**
+ * Given errors from a search query this will return an array of strings derived from the errors.
+ * @param errors The errors to derive the strings from
+ */
+export const createErrorsFromShard = ({ errors }: { errors: ShardError[] }): string[] => {
+  return errors.map((error) => {
+    return `reason: ${error.reason.reason}, type: ${error.reason.caused_by.type}, caused by: ${error.reason.caused_by.reason}`;
+  });
+};
+
+export const createSearchAfterReturnTypeFromResponse = ({
+  searchResult,
+}: {
+  searchResult: SignalSearchResponse;
+}): SearchAfterAndBulkCreateReturnType => {
+  return createSearchAfterReturnType({
+    success: searchResult._shards.failed === 0,
+    lastLookBackDate:
+      searchResult.hits.hits.length > 0
+        ? new Date(searchResult.hits.hits[searchResult.hits.hits.length - 1]?._source['@timestamp'])
+        : undefined,
+  });
+};
+
+export const createSearchAfterReturnType = ({
+  success,
+  searchAfterTimes,
+  bulkCreateTimes,
+  lastLookBackDate,
+  createdSignalsCount,
+  errors,
+}: {
+  success?: boolean | undefined;
+  searchAfterTimes?: string[] | undefined;
+  bulkCreateTimes?: string[] | undefined;
+  lastLookBackDate?: Date | undefined;
+  createdSignalsCount?: number | undefined;
+  errors?: string[] | undefined;
+} = {}): SearchAfterAndBulkCreateReturnType => {
+  return {
+    success: success ?? true,
+    searchAfterTimes: searchAfterTimes ?? [],
+    bulkCreateTimes: bulkCreateTimes ?? [],
+    lastLookBackDate: lastLookBackDate ?? null,
+    createdSignalsCount: createdSignalsCount ?? 0,
+    errors: errors ?? [],
+  };
+};
+
+export const mergeSearchAfterReturnTypeFromResponse = ({
+  searchResult,
+  prev,
+  next,
+}: {
+  searchResult: SignalSearchResponse;
+  prev: SearchAfterAndBulkCreateReturnType;
+  next: SearchAfterAndBulkCreateReturnType;
+}): SearchAfterAndBulkCreateReturnType => {
+  const searchReturn = createSearchAfterReturnType({
+    success: searchResult._shards.failed === 0,
+    lastLookBackDate:
+      searchResult.hits.hits.length > 0
+        ? new Date(searchResult.hits.hits[searchResult.hits.hits.length - 1]?._source['@timestamp'])
+        : undefined,
+  });
+  const partialMerge = mergeSearchAfterAndBulkCreate({
```

Comment:
I have a suggestion to normalize these merging functions, but I realize it might cause a bunch of churn so feel free to ignore it. The basic idea is to decompose into the following functions:

* function to merge two `SearchAfterAndBulkCreateReturnType`s into one (e.g.`(a: A, b: A) => A`)
* function to reduce n `SearchAfterAndBulkCreateReturnType`s via the above function, using `createSearchAfterReturnType()` as the starting value

That would eliminate the need for these prev/next parameters and the partialMerge stuff. Again, just a suggestion, take it or leave it!